### PR TITLE
Starting to extract creation of DFG nodes into separate pass

### PR DESCRIPTION
**Pull request by KuechA**
This PR introduces a new pass, the `DFGPass` which tries to consolidate the creation of data flow graph edges into one pass. Previously, this was done in a lot of places, e.g., in the setters of the individual nodes. We are incrementally moving this into the new pass.

### CODE REVIEW

**Comment by KuechA** [comment]
The new Pass somehow gets stuck in *some* go tests.

**Comment by oxisto** [comment]
> The new Pass somehow gets stuck in _some_ go tests.

testDFG and testArrayCompositeLiteral in GoLanguageFrontendTest get stuck

**Comment by oxisto** [comment]
> > The new Pass somehow gets stuck in _some_ go tests.
> 
> testDFG and testArrayCompositeLiteral in GoLanguageFrontendTest get stuck

It gets stuck in the following loop:

https://github.com/Fraunhofer-AISEC/cpg/blob/05a447e356585cfffe5f8e9673521de691fbf0d0/cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt#L267-L269

What's the point of the loop?

**Comment by KuechA** [comment]
The purpose of that loop was to remove already existing DFG edges. For `CallExpression`s, the current implementation of the `CallResolver` makes use of DFG edges. However, after actually resolving the call target in the `CallResolver`, we can update the DFG edges of that `CallExpression` to enable flows to arguments and from the return value. To avoid superfluous and no longer correct DFG edges, we remove the existing `prevDFG` of the `CallExpression` and also have to remove the expression from the respective previous node's `nextDFG` set.

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt [hunk 1]
```diff
@@ -0,0 +1,292 @@
...
+import de.fraunhofer.aisec.cpg.graph.declarations.VariableDeclaration
+import de.fraunhofer.aisec.cpg.graph.statements.CompoundStatement
+import de.fraunhofer.aisec.cpg.graph.statements.ReturnStatement
+import de.fraunhofer.aisec.cpg.graph.statements.expressions.*
+import de.fraunhofer.aisec.cpg.helpers.SubgraphWalker.IterativeGraphWalker
+import de.fraunhofer.aisec.cpg.helpers.Util
+
+/** Adds the DFG edges for various types of nodes. */
+@DependsOn(VariableUsageResolver::class)
+class UnresolvedDFGPass : Pass() {
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt:43]: Just name it `DFGPass`? or `DataflowGraphPass`

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/FunctionPointerCallResolver.kt [hunk 2]
```diff
@@ -152,6 +154,7 @@ class FunctionPointerCallResolver : Pass() {
         }
 
         call.invokes = invocationCandidates
+        UnresolvedDFGPass().handleCallExpression(call, inferDfgForUnresolvedCalls)
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/FunctionPointerCallResolver.kt:157]: A comment here would be good why we are "manually" invoking the pass here.

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt [hunk 3]
```diff
@@ -0,0 +1,292 @@
...
+            // Other
+            is Assignment -> handleAssignment(node)
+        }
+    }
+
+    /**
+     * Adds the DFG edge for a [FunctionDeclaration]. The data flow from the return statement(s) to
+     * the function.
+     *
+     * TODO: Does not seem to work (yet)! Some function pointer tests fail
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt:88]: Still TODO?

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt [hunk 4]
```diff
@@ -0,0 +1,292 @@
...
+        } else if (node.body is CompoundStatement) {
+            (node.body as CompoundStatement)
+                .statements
+                .filterIsInstance<ReturnStatement>()
+                .forEach { node.addPrevDFG(it) }
+        }
+    }
+
+    /**
+     * Adds the DFG edge for a [FieldDeclaration]. The data flow from the initializer to the field.
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt:110]: flows? similar spelling in other functions.

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt [hunk 5]
```diff
@@ -0,0 +1,292 @@
...
+     */
+    private fun handleKeyValueExpression(node: KeyValueExpression) {
+        node.value?.let { node.addPrevDFG(it) }
+    }
+
+    /**
+     * Adds the DFG edges for an [InitializerListExpression]. All values in the initializer flow to
+     * this expression.
+     *
+     * TODO: This change seems to have performance issues!
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt:161]: Still TODO?

File: cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt [hunk 6]
```diff
@@ -0,0 +1,292 @@
...
+        node.expressions.lastOrNull()?.let { node.addPrevDFG(it) }
+    }
+
+    /**
+     * Adds the DFG edges to a [DeclaredReferenceExpression] as follows:
+     * - If the variable is written to, data flows from this node to the variable declaration.
+     * - If the variable is read from, data flows from the variable declaration to this node.
+     * - For a combined read and write, both edges for data flows are added.
+     *
+     * TODO: Doesn't seem to work (yet)! Some function pointer tests fail
```
Comment (oxisto) [inline cpg-core/src/main/java/de/fraunhofer/aisec/cpg/passes/UnresolvedDFGPass.kt:184]: Still TODO?

**Review by oxisto — APPROVED** [review]

**Pull request merged**